File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int n = s.length();
        int m = t.length();
        vector<vector<unsigned long long>> dp(n+1,vector<unsigned long long> (m+1,-1));
        for(int i = 0;i<=m;i++) dp[0][i] = 0;
        for(int j = 0;j<=n;j++) dp[j][0] = 1;
        for(int i = 1;i<=n;i++){
            for(int j =1;j<=m;j++){
                dp[i][j] = dp[i-1][j];
                if(s[i-1]==t[j-1]) dp[i][j]+=dp[i-1][j-1];
            }
        }
        return dp[n][m];
    }
};

File: 0115-distinct-subsequences/0115-distinct-subsequences.py
from functools import lru_cache

class Solution:
    def numDistinct(self, s: str, t: str) -> int:

        if len(s) < len(t):
            return 0

        @lru_cache(None)
        def recurse(i: int, j: int) -> int:

            if j == len(t):
                return 1

            if i == len(s):
                return 0

            if s[i] == t[j]:
                take = recurse(i + 1, j + 1)
                leave = recurse(i + 1, j)
                return take + leave

            return recurse(i + 1, j)

        return recurse(0, 0)